include/roboteam_ai/world/Ball.h

```diff
@@ -15,41 +11,130 @@ namespace world {
+        /* The threshold used for the filtered velocity, if it exceeds this value then we use the expectedVelocity
+        instead as estimation for the filtered velocity. More information can be found in the comment in the
+        filterBallVelocity method of the Ball.cpp file. */
+        const float MAXIMUM_FILTER_VELOCITY = 100.0;
+
     public:
         using BallPtr = std::shared_ptr<Ball>;
         using RobotPtr = std::shared_ptr<Robot>;
 
     private:
-        bool ballInAir = false;
-        bool collidesNow = false;
-        bool kickedNow = false;
-        bool dribbledNow = false;
-        double lastKickVel = 0;
-        int ballStraightTicks = 0;
-        Vector2 ballStillPosition = Vector2();
-
-        void updateDribbling(const Ball &oldBall, const WorldData &worldData);
-        Robot* getDribblingRobot(const std::vector<RobotPtr> &robots, double maxDribbleRange);
-        void updateBallModel(const Ball &oldBall, const WorldData &worldData);
-        void updateBallPosition(const Ball &oldBall, const WorldData &worldData);
-        void updateExpectedPositionWhereBallIsStill(const Ball &oldBall, const WorldData &worldData);
+        Vector2 expectedPosition = Vector2(); // The expected position where this ball is.
+        const Vector2 expectedVelocity = Vector2(); // The expected velocity of the ball.
+        /* If the ball is visible by any camera. This value is true if the ball is visible by any camera and false
+         * otherwise */
+        bool visibleByAnyCamera = false;
+        // The position where the ball is expected to end (lay still) after following his path.
+        Vector2 expectedBallEndPosition = Vector2();
+        // The velocity but then adjusted to determine a more realistic expectedBallEndPosition
+        Vector2 filteredVelocity = Vector2();
+
+        /**
+         * When the position of the ball is not yet initialized, but was previously near to a robot then initialize
+         * the position of the ball to a position close to that robot.
+         */
+        void initBallAtRobotPosition(const Ball &oldBall, const WorldData &worldData);
+
+        /**
+         * Adds a moving average over the Kalman filter to make the ball-velocity more stable (this method could be
+         * moved to rtt_world).
+        */
+        void filterBallVelocity(Ball &oldBall, const WorldData &worldData);
+
+        /**
+         * Update the expected position where the ball will end (lay still) after following his path.
+         */
+        void updateExpectedBallEndPosition(const Ball &oldBall, const WorldData &worldData);
+
+        /**
+         * When the position of the ball is close to a robot and not visible by camera then update the ball according
+         * to the movement and rotation of that robot (which happens during dribbling).
+         */
+        void updateBallAtRobotPosition(const Ball &oldBall, const WorldData &worldData);
 
     public:
+        static bool exists;
```

Comment:
Should this be static?

Reply:
Yes. Don't ask. As I mentioned above, this is an ugly hack that should be remove, but it requires changes in roboteam_world/roboteam_proto to fix aswell.